builds/jenkins-build-sls-app/Jenkinsfile

```diff
@@ -0,0 +1,1062 @@
+        events.sendFailureEvent('UPDATE_DEPLOYMENT', ex.getMessage(), environmentDeploymentMetadata.generateDeploymentMap("failed", environment_logical_id, gitCommitHash), environment_logical_id)
+        error ex.getMessage()
+      }
+      events.sendCompletedEvent('VALIDATE_PRE_BUILD_CONF', 'pre-build validation completed', context_map, environment_logical_id)
+    }
+
+    if (config['domain'] && config['domain'] != 'jazz' && configLoader.CODE_QUALITY.SONAR.ENABLE_SONAR == "true") {
+      stage('Code Quality Check') {
+        events.sendStartedEvent('CODE_QUALITY_CHECK', 'code quality check starts', context_map, environment_logical_id)
+        try {
+          runValidation(runtime)
+          sonarModule.doAnalysis()
+        } catch (ex) {
+          events.sendFailureEvent('CODE_QUALITY_CHECK', ex.getMessage(), context_map, environment_logical_id)
+          events.sendFailureEvent('UPDATE_DEPLOYMENT', ex.getMessage(), environmentDeploymentMetadata.generateDeploymentMap("failed", environment_logical_id, gitCommitHash), environment_logical_id)
+          error ex.getMessage()
+        }
+        events.sendCompletedEvent('CODE_QUALITY_CHECK', 'code quality check completed', context_map, environment_logical_id)
+      }
+    }
+
+    stage('Build') {
+      events.sendStartedEvent('BUILD', 'build starts', context_map, environment_logical_id)
+      try {
+        buildLambda(runtime, repo_name)
+      } catch (ex) {
+        events.sendFailureEvent('BUILD', ex.getMessage(), context_map, environment_logical_id)
+        events.sendFailureEvent('UPDATE_DEPLOYMENT', ex.getMessage(), environmentDeploymentMetadata.generateDeploymentMap("failed", environment_logical_id, gitCommitHash), environment_logical_id)
+        error ex.getMessage()
+      }
+      events.sendCompletedEvent('BUILD', 'build completed', context_map, environment_logical_id)
+    }
+
+    def env_key
+    if (branch == "master") {
+      env_key = "PROD"
+    } else {
+      env_key = "DEV"
+    }
+
+    stage('Package') {
+      echo 'The packaging state'
+      withCredentials([
+        [$class: 'AmazonWebServicesCredentialsBinding', accessKeyVariable: 'AWS_ACCESS_KEY_ID', credentialsId: configLoader.AWS_CREDENTIAL_ID, secretKeyVariable: 'AWS_SECRET_ACCESS_KEY']
+      ]) {
+
+        def randomString = utilModule.generateRequestId();
+        def credsId = "jazz-${randomString}";
+        sh "aws configure set profile.${credsId}.region ${configLoader.AWS.REGION}"
+        sh "aws configure set profile.${credsId}.aws_access_key_id $AWS_ACCESS_KEY_ID"
+        sh "aws configure set profile.${credsId}.aws_secret_access_key $AWS_SECRET_ACCESS_KEY"
+
+        try {
+          sh "serverless package --stage ${environment_logical_id} -v > output.log"
+        } catch(ex) {
+          def outputLog = readFile('output.log').trim()
+          echo "serverless packaging log $outputLog"
+          throw ex
```

Comment:
Where do we catch this error? We'll need to call `events.sendFailureEvent('$EVENT_NAME....)` in case of any error in the workflow to notify Jazz & mark this deployment as failed. 

Please send similar failure events wherever you see in error.

Reply:
Well, here the 'throw ex' is equal to 'error ex.getMessage()' that is used throughout the code right now. The 'error message' expression also raises the exception according to https://jenkins.io/doc/pipeline/steps/workflow-basic-steps/ where it says following: 'Signals an error. Useful if you want to conditionally abort some part of your program. You can also just throw new Exception(), but this step will avoid printing a stack trace'. However for the sake of unification I changed this part. Resolved with https://github.com/tmobile/jazz/pull/829/commits/ed93653c678f74bdc629c7932e2d8cc043e9e85f